runtime/vm/exceptiondescribe.c

```diff
@@ -343,10 +343,25 @@ iterateStackTrace(J9VMThread * vmThread, j9object_t* exception, callback_func_t
 					pruneConstructors = FALSE;
 #endif
 					romClass = findROMClassFromPC(vmThread, methodPC, &classLoader);
-					if(romClass) {
-						romMethod = findROMMethodInROMClass(vmThread, romClass, methodPC);
-						if (romMethod != NULL) {
-							methodPC -= (UDATA) J9_BYTECODE_START_FROM_ROM_METHOD(romMethod);
+					if (NULL != romClass) {
+						J9UTF8 const *utfClassName = J9ROMCLASS_CLASSNAME(romClass);
+
+						ramClass = peekClassHashTable(vmThread, classLoader, J9UTF8_DATA(utfClassName), J9UTF8_LENGTH(utfClassName));
+						if (NULL != ramClass) {
+							U_32 i = 0;
+							J9Method *methods = ramClass->ramMethods;
+							for (i = 0; i < romClass->romMethodCount; ++i) {
+								J9ROMMethod *possibleMethod = J9_ROM_METHOD_FROM_RAM_METHOD(&methods[i]);
+
+								/* Note that we cannot use `J9_BYTECODE_START_FROM_ROM_METHOD` here because native method PCs
+								 * point to the start of the J9ROMMethod data structure
+								 */
+								if ((methodPC >= (UDATA)possibleMethod) && (methodPC < (UDATA)J9_BYTECODE_END_FROM_ROM_METHOD(possibleMethod))) {
+									romMethod = possibleMethod;
+									methodPC -= (UDATA)J9_BYTECODE_START_FROM_ROM_METHOD(romMethod);
+									break;
+								}
+							}
```

Comment:
I'm not sure if the PR testing will catch it but there seems to almost certainly be a problem with the this code from 
da33dfb. It seems when this code is enabled we hit the assert in [1] for some reason. Another (internal) test seems to fail where I can see a bunch of:

```
at <unknown class>.<unknown method>(Unknown Source)
```

In the backtrace which looks incorrect. Previously the code on line 359 was:

```
if ((methodPC >= (UDATA)J9_BYTECODE_START_FROM_ROM_METHOD(possibleMethod)) && (methodPC < (UDATA)J9_BYTECODE_END_FROM_ROM_METHOD(possibleMethod))) {
	...
}
```

Which did not work because it broke the test case in [2]. It was returning the following:

```
"main" Id=1 RUNNABLE
        at <unknown class>.<unknown method>(Unknown Source)
        at com.ibm.java.lang.management.internal.ThreadMXBeanImpl.getThreadInfo(ThreadMXBeanImpl.java:266)
```

instead of:

```
"main" prio=5 Id=1 RUNNABLE
        at app//com.ibm.java.lang.management.internal.ThreadMXBeanImpl.getMultiThreadInfoImpl(Native Method)
        at app//com.ibm.java.lang.management.internal.ThreadMXBeanImpl.getThreadInfo(ThreadMXBeanImpl.java:266)
```

The reason seems to be that native methods have a different `methodPC`. This is also confirmed by looking at the implementation of `findROMMethodInROMClass` [3] which also does not use `J9_BYTECODE_START_FROM_ROM_METHOD` and compares the `methodPC` directly against the `J9ROMMethod*`.

There seems to be another bug here now which is causing the assert [2] to be triggered and other backtrace strings not to be generated.

[1] https://github.com/eclipse/openj9/blob/60ed51cffd39410084955081e3f37667ace61717/runtime/jcl/common/jclexception.c#L263
[2] https://github.com/eclipse/openj9/blob/60ed51cffd39410084955081e3f37667ace61717/test/functional/JLM_Tests/src/org/openj9/test/java/lang/management/TestManagementFactory.java#L1306-L1307
[3] https://github.com/eclipse/openj9/blob/60ed51cffd39410084955081e3f37667ace61717/runtime/vm/findmethod.c#L60-L63

Reply:
@DanHeidinga was wondering if you could have a glance and maybe your expertise may be able to eyeball what the issue might be.

Reply:
The code also needs to handle redefined classes.  If we don't find the method in the `ramClass` returned by `peekClassHashTable(...)`, then we need to check its `ramClass->replacedClass` linked list to see if the method is found in one of the previous versions of this class.

The scenario would be a stacktrace is collected and then one of the classes in the stacktrace is redefined, which would lead to a new ROMClass being created.  The PC stored in the stacktrace wouldn't map to the current version of the J9Class.

Addressing this should cover some (all?) of the `<unknown class>.<unknown method>(Unknown Source)` cases

Reply:
https://github.com/eclipse/openj9/blob/60ed51cffd39410084955081e3f37667ace61717/runtime/jcl/common/jclexception.c#L254

Can you also assert immediately after this line?  It would help to determine which case has the error.  I assume it must be this one but ....

Is it always the same test that asserts?  If so, we can add code that explicitly checks for null and then printfs the `result`, `element`, & `currentIndex - 1` and then we can look at the values in DDR to see if there's a logical code path that explains them being null

Reply:
I tried the `replacedClass` fix but it did not fix the problems observed on the internal tests:
https://github.com/fjeremic/openj9/commit/24566add8fb003b49500670e14fc7cb5ba960719

It is always the same test that fails. There are two classes of errors. One is the assert triggering and the other is some weird NPE error where the backtrace looks like this:

```
[ERROR   ] CNTR0020E: EJB threw an unexpected (non-declared) exception during invocation of method "getAccountData" on bean "BeanId(daytrader3#dt-ejb.jar#TradeSLSBBean, null)". Exception data: java.lang.NullPointerException
	at com.ibm.websphere.samples.daytrader.ejb3.TradeSLSBBean.getAccountData(TradeSLSBBean.java:479)
	at com.ibm.websphere.samples.daytrader.ejb3.EJSLocal0SLTradeSLSBBean_13a83d61.getAccountData(EJSLocal0SLTradeSLSBBean_13a83d61.java)
	at com.ibm.websphere.samples.daytrader.web.jsf.AccountBean.userLookup(AccountBean.java:68)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:90)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
	at java.lang.reflect.Method.invoke(Method.java:508)
	at <unknown class>.<unknown method>(Unknown Source)
	at <unknown class>.<unknown method>(Unknown Source)
	at <unknown class>.<unknown method>(Unknown Source)
	at <unknown class>.<unknown method>(Unknown Source)
	at <unknown class>.<unknown method>(Unknown Source)
	at <unknown class>.<unknown method>(Unknown Source)
	at <unknown class>.<unknown method>(Unknown Source)
...
	at <unknown class>.<unknown method>(Unknown Source)
	at <unknown class>.<unknown method>(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1160)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.lang.Thread.run(Thread.java:820)
```

I'll get a build going with printfs and see where we can get.

Reply:
I dug a bit deeper with DDR to find out why we asserted. We do indeed enter the `if (previousFileName == fileName)` block and we fail to extract a valid string. Looking at the value of `result` we see the following array:

```
> !j9indexableobject 0x000003FF9AD50980 0 38
!J9IndexableObject 0x000003FF9AD50980 {
    struct J9Class* clazz = !j9arrayclass 0x3FFB019A000   // [L
    Object flags = 0x00000000;
    U_32 size = 0x00000026;
        [0] = !fj9object 0x3ff9ad50ac0 = !j9object 0x3ff9ad50ac0
        [1] = !fj9object 0x3ff9ad50af8 = !j9object 0x3ff9ad50af8
        [2] = !fj9object 0x3ff9ad50b30 = !j9object 0x3ff9ad50b30
        [3] = !fj9object 0x3ff9ad50b68 = !j9object 0x3ff9ad50b68
        [4] = !fj9object 0x3ff9ad50ba0 = !j9object 0x3ff9ad50ba0
        [5] = !fj9object 0x3ff9ad50bd8 = !j9object 0x3ff9ad50bd8
        [6] = !fj9object 0x3ff9ad50c10 = !j9object 0x3ff9ad50c10
        [7] = !fj9object 0x3ff9ad50c48 = !j9object 0x3ff9ad50c48
        [8] = !fj9object 0x3ff9ad50c80 = !j9object 0x3ff9ad50c80
        [9] = !fj9object 0x3ff9ad50cb8 = !j9object 0x3ff9ad50cb8
        [10] = !fj9object 0x3ff9ad50cf0 = !j9object 0x3ff9ad50cf0
        [11] = !fj9object 0x3ff9ad50d28 = !j9object 0x3ff9ad50d28
        [12] = !fj9object 0x3ff9ad50d60 = !j9object 0x3ff9ad50d60
        [13] = !fj9object 0x3ff9ad50d98 = !j9object 0x3ff9ad50d98
        [14] = !fj9object 0x3ff9ad50dd0 = !j9object 0x3ff9ad50dd0
        [15] = !fj9object 0x3ff9ad50e08 = !j9object 0x3ff9ad50e08
        [16] = !fj9object 0x3ff9ad50e40 = !j9object 0x3ff9ad50e40
        [17] = !fj9object 0x3ff9ad50e78 = !j9object 0x3ff9ad50e78
        [18] = !fj9object 0x3ff9ad50eb0 = !j9object 0x3ff9ad50eb0
        [19] = !fj9object 0x3ff9ad50ee8 = !j9object 0x3ff9ad50ee8
        [20] = !fj9object 0x3ff9ad50f20 = !j9object 0x3ff9ad50f20
        [21] = !fj9object 0x3ff9ad50f58 = !j9object 0x3ff9ad50f58
        [22] = !fj9object 0x3ff9ad50f90 = !j9object 0x3ff9ad50f90
        [23] = !fj9object 0x3ff9ad50fc8 = !j9object 0x3ff9ad50fc8
        [24] = !fj9object 0x3ff9ad51000 = !j9object 0x3ff9ad51000
        [25] = !fj9object 0x3ff9ad51038 = !j9object 0x3ff9ad51038
        [26] = !fj9object 0x3ff9ad51070 = !j9object 0x3ff9ad51070
        [27] = !fj9object 0x3ff9ad510a8 = !j9object 0x3ff9ad510a8
        [28] = !fj9object 0x3ff9ad510e0 = !j9object 0x3ff9ad510e0
        [29] = !fj9object 0x3ff9ad51118 = !j9object 0x3ff9ad51118
        [30] = null
        [31] = null
        [32] = null
        [33] = null
        [34] = null
        [35] = null
        [36] = null
        [37] = null
}
```

What is interesting here is that only the first index holds a valid `StackTraceElement`:

```
> !fj9object 0x3ff9ad50ac0
!J9Object 0x000003FF9AD50AC0 {
        struct J9Class* clazz = !j9class 0x3FFB0198A00 // java/lang/StackTraceElement
        Object flags = 0x00000000;
        J lockword = 0x0000000000000000 (offset = 0) (java/lang/Object) <hidden>
        Ljava/lang/String; declaringClass = !fj9object 0x3ff8ee72b30 (offset = 8) (java/lang/StackTraceElement)
        Ljava/lang/String; methodName = !fj9object 0x3ff8f84f478 (offset = 16) (java/lang/StackTraceElement)
        Ljava/lang/String; fileName = !fj9object 0x3ff8ee727f0 (offset = 24) (java/lang/StackTraceElement)
        I lineNumber = 0x0000048A (offset = 40) (java/lang/StackTraceElement)
        Ljava/lang/Object; source = !fj9object 0x0 (offset = 32) (java/lang/StackTraceElement)
}
> !fj9object 0x3ff9ad50af8
!J9Object 0x000003FF9AD50AF8 {
        struct J9Class* clazz = !j9class 0x3FFB0198A00 // java/lang/StackTraceElement
        Object flags = 0x00000000;
        J lockword = 0x0000000000000000 (offset = 0) (java/lang/Object) <hidden>
        Ljava/lang/String; declaringClass = !fj9object 0x0 (offset = 8) (java/lang/StackTraceElement)
        Ljava/lang/String; methodName = !fj9object 0x0 (offset = 16) (java/lang/StackTraceElement)
        Ljava/lang/String; fileName = !fj9object 0x0 (offset = 24) (java/lang/StackTraceElement)
        I lineNumber = 0x00000000 (offset = 40) (java/lang/StackTraceElement)
        Ljava/lang/Object; source = !fj9object 0x0 (offset = 32) (java/lang/StackTraceElement)
}
```

The rest of them are all valid objects but have `NULL` for `declaringClass`, `methodName`, and `fileName`. The `printf` printed the following:

```
Filip: result = 0x3ff9ad50980, element = 0x3ff9ad510e0, currentIndex - 1 = 28, romClass = 0x3ff5f34d260, romMethod = 0x3ff5f34da20, fileName = 0x3ff5f34e3d0, lineNumber = 236, classLoader = 0x3ffb0070990, ramClass = 0x3ffb0570300
```

Which means we compared `previousFileName == fileName` and found it to be true on the 28th index. However looking at the 27th object in the array:

```
> !fj9object 0x3ff9ad510a8
!J9Object 0x000003FF9AD510A8 {
        struct J9Class* clazz = !j9class 0x3FFB0198A00 // java/lang/StackTraceElement
        Object flags = 0x00000000;
        J lockword = 0x0000000000000000 (offset = 0) (java/lang/Object) <hidden>
        Ljava/lang/String; declaringClass = !fj9object 0x0 (offset = 8) (java/lang/StackTraceElement)
        Ljava/lang/String; methodName = !fj9object 0x0 (offset = 16) (java/lang/StackTraceElement)
        Ljava/lang/String; fileName = !fj9object 0x0 (offset = 24) (java/lang/StackTraceElement)
        I lineNumber = 0x00000000 (offset = 40) (java/lang/StackTraceElement)
        Ljava/lang/Object; source = !fj9object 0x0 (offset = 32) (java/lang/StackTraceElement)
}
```

The `fileName` is `NULL`. I'm sure having all these `NULL` values within these objects is not normal. So I set out for another experiment which fixes the problem. I'll describe that in the following comment.

Reply:
Using the branch with the `printf` to gather more information I tried the following change:
https://github.com/fjeremic/openj9/commit/87f058fd670563f89ad9f1a85f961e801b22a35b

After trying to iterate `J9Method*`, it turns out we were sometimes unable to find the corresponding `J9ROMMethod*`, because the `peekClassHashTable` returned a `NULL` class pointer. So we were proceeding with a bogus value for `romMethod` and not extracting proper information.

Adding in the fallback path on lines 369-374:
https://github.com/fjeremic/openj9/commit/87f058fd670563f89ad9f1a85f961e801b22a35b#diff-32ef01e828ed7705ed9d726657f20b8dR369-R374

fixes the issue and we are now passing all the tests. It seems that quite frequently we will get back a `NULL` result from `peekClassHashTable`. I'm unsure as to why that is the case. @DanHeidinga do you may have any ideas? Should we go forward with this new implementation which does a fallback if we can't find the ROM method because we have a `NULL` class pointer?

Reply:
@DanHeidinga apologies for the spam. Tagging you again just in case you missed the previous question.

Reply:
@fjeremic thanks for the ping on this.  I had missed the previously question while at DevNexus.

I'm trying to understand the case that's causing `peekClassHashTable` to return NULL and the only thing I can think of is that the class has been unloaded but we're keeping the ram/rom classes around to describe the "dark matter" on the heap for the GC.  This would mean the classes can't be found from the ClassHashTable but would still be findable by walking the Segments.

If you have a core from the problematic case, can you check the classloader being used and see if it has the `J9_GC_CLASS_LOADER_DEAD` bit set in the `gcFlags` field?  ie: `(J9_GC_CLASS_LOADER_DEAD == (classLoader->gcFlags & J9_GC_CLASS_LOADER_DEAD) `

Or if the class has the `J9AccClassDying` bit set in the `J9CLASS_FLAGS(clazz)` - ie: `(J9AccClassDying == (J9CLASS_FLAGS(clazz) & J9AccClassDying))`?

Reply:
I checked the core file which the previous DDR comments were based on:

```
> !j9classloader 0x3ffb0070990
J9ClassLoader at 0x3ffb0070990 {
  Fields for J9ClassLoader:
  ...
          0x60: UDATA flags = 0x0000000000000118 (280)

> !j9class 0x3ffb0570300
J9Class at 0x3ffb0570300 {
  Fields for J9Class:
  ...
        0x18: UDATA classDepthAndFlags = 0x00000000021E0002 (35520514)
        0x20: U32 classDepthWithFlags = 0x00000000 (0)
        0x24: U32 classFlags = 0x00000000 (0)
```

It seems the class loader is not dead (0x2 flag) and there are no class flags. I also added some print statements to see which classes we couldn't find via `peekClassHashTable` using the following commit:
https://github.com/fjeremic/openj9/commit/2d449133ebef647c942d2a2471b2b10c88ee3e6b

Here is a snapshot of the output on one of the problematic tests:

```
j> 12:54:30 Filip: counterFound = 9, class name = java/lang/Thread
j> 12:54:30 Filip: counter **NOT** found = 95, ramClassFound = 0, loaderDead = 0, gcFlags = 0, class name = com/ibm/test/binaryData/TestBinaryConverters
j> 12:54:30 Filip: counterFound = 10, class name = sun/reflect/NativeMethodAccessorImpl
j> 12:54:30 Filip: counter **NOT** found = 96, ramClassFound = 0, loaderDead = 0, gcFlags = 0, class name = org/junit/runners/ParentRunner$3
j> 12:54:30 Filip: counter **NOT** found = 97, ramClassFound = 0, loaderDead = 0, gcFlags = 0, class name = org/junit/runners/ParentRunner$1
j> 12:54:30 Filip: counter **NOT** found = 98, ramClassFound = 0, loaderDead = 0, gcFlags = 0, class name = org/junit/runners/ParentRunner
j> 12:54:30 Filip: counter **NOT** found = 99, ramClassFound = 0, loaderDead = 0, gcFlags = 0, class name = org/junit/runners/ParentRunner
j> 12:54:30 Filip: counter **NOT** found = 100, ramClassFound = 0, loaderDead = 0, gcFlags = 0, class name = org/junit/runners/ParentRunner$2
j> 12:54:30 Filip: counter **NOT** found = 101, ramClassFound = 0, loaderDead = 0, gcFlags = 0, class name = org/junit/internal/runners/statements/RunBefores
j> 12:54:30 Filip: counter **NOT** found = 102, ramClassFound = 0, loaderDead = 0, gcFlags = 0, class name = org/junit/runners/ParentRunner
j> 12:54:30 Filip: counter **NOT** found = 103, ramClassFound = 0, loaderDead = 0, gcFlags = 0, class name = org/junit/runners/Suite
j> 12:54:30 Filip: counter **NOT** found = 104, ramClassFound = 0, loaderDead = 0, gcFlags = 0, class name = org/junit/runners/Suite
j> 12:54:30 Filip: counter **NOT** found = 105, ramClassFound = 0, loaderDead = 0, gcFlags = 0, class name = org/junit/runners/ParentRunner$3
j> 12:54:30 Filip: counter **NOT** found = 106, ramClassFound = 0, loaderDead = 0, gcFlags = 0, class name = org/junit/runners/ParentRunner$1
j> 12:54:30 Filip: counter **NOT** found = 107, ramClassFound = 0, loaderDead = 0, gcFlags = 0, class name = org/junit/runners/ParentRunner
j> 12:54:30 Filip: counter **NOT** found = 108, ramClassFound = 0, loaderDead = 0, gcFlags = 0, class name = org/junit/runners/ParentRunner
j> 12:54:30 Filip: counter **NOT** found = 109, ramClassFound = 0, loaderDead = 0, gcFlags = 0, class name = org/junit/runners/ParentRunner$2
j> 12:54:30 Filip: counter **NOT** found = 110, ramClassFound = 0, loaderDead = 0, gcFlags = 0, class name = org/junit/runners/ParentRunner
j> 12:54:30 Filip: counter **NOT** found = 111, ramClassFound = 0, loaderDead = 0, gcFlags = 0, class name = org/junit/runner/JUnitCore
j> 12:54:30 Filip: counter **NOT** found = 112, ramClassFound = 0, loaderDead = 0, gcFlags = 0, class name = org/junit/runner/JUnitCore
j> 12:54:30 Filip: counter **NOT** found = 113, ramClassFound = 0, loaderDead = 0, gcFlags = 0, class name = com/ibm/jtc/test/jltf/workunitadapters/JUnit4WorkUnit
j> 12:54:30 Filip: counter **NOT** found = 114, ramClassFound = 0, loaderDead = 0, gcFlags = 0, class name = com/ibm/jtc/test/jltf/threads/WorkUnitRunner
j> 12:54:30 Filip: counter **NOT** found = 115, ramClassFound = 0, loaderDead = 0, gcFlags = 0, class name = com/ibm/jtc/test/jltf/threads/pooling/PooledRunnable
j> 12:54:30 Filip: counter **NOT** found = 116, ramClassFound = 0, loaderDead = 0, gcFlags = 0, class name = com/ibm/jtc/test/jltf/threads/pooling/PooledRunnable
j> 12:54:30 Filip: counter **NOT** found = 117, ramClassFound = 0, loaderDead = 0, gcFlags = 0, class name = com/ibm/jtc/test/jltf/threads/PooledLangThreadFactory$JavaLangPooledThread
```

The classes which were not found via `peekClassHashTable` by their name definitely look like short lived classes. It also seems they are classes loaded by the application class loader. I could not test the `J9AccClassDying` flag as I don't have a `J9Class*` in hand.

Reply:
I also tried iterating the `classLoader->classSegments` to look for the `J9Class*` for the `romClass` we have in hand via:

```
                                        J9MemorySegment *segment = classLoader->classSegments;
                                        while (NULL != segment) {
                                                if (MEMORY_TYPE_RAM_CLASS == (segment->type & MEMORY_TYPE_RAM_CLASS)) {
                                                        J9Class* segClass = *((J9Class**)segment->heapBase);
                                                        while (NULL != segClass) {
                                                                if (segClass->romClass == romClass) {
                                                                        fprintf(stderr, "Filip: Found RAMClass = %p\n", segClass);
                                                                }
                                                                segClass = segClass->nextClassInSegment;
                                                        }
                                                }
                                                segment = segment->nextSegmentInClassLoader;
                                        }
```

And it was not printing anything. I assume there has to be a `J9Class*` for every `J9ROMClass*` we have in hand, right? Not sure if the above algorithm covers all cases.

Reply:
@fjeremic J9Classloader has both a `flags` and a `gcFlags` field.  Can you dump the `gcFlags` field (or the whole structure) for `!j9classloader 0x3ffb0070990`?

Reply:
```
> !j9classloader 0x3ffb0070990
J9ClassLoader at 0x3ffb0070990 {
  Fields for J9ClassLoader:
        0x0: struct J9Pool * sharedLibraries = !j9pool 0x0000000000000000
        0x8: struct J9HashTable * classHashTable = !j9hashtable 0x000003FFB028C9A0
        0x10: struct J9HashTable * romClassOrphansHashTable = !j9hashtable 0x0000000000000000
        0x18: j9object_t classLoaderObject = !j9object 0x000003FF8E6E64C0 // sun/misc/Launcher$AppClassLoader
        0x20: struct J9ClassPathEntry * classPathEntries = !j9classpathentry 0x000003FFB04CC210
        0x28: UDATA classPathEntryCount = 0x0000000000000000 (0)
        0x30: struct J9ClassLoader * unloadLink = !j9classloader 0x0000000000000000
        0x38: struct J9ClassLoader * gcLinkNext = !j9classloader 0x0000000000000000
        0x40: struct J9ClassLoader * gcLinkPrevious = !j9classloader 0x0000000000000000
        0x48: UDATA gcFlags = 0x0000000000000000 (0)
        0x50: struct J9VMThread * gcThreadNotification = !j9vmthread 0x0000000000000000
        0x58: struct J9Pool * jniIDs = !j9pool 0x000003FFB028CA80
        0x60: UDATA flags = 0x0000000000000118 (280)
        0x68: struct J9JITExceptionTable * jitMetaDataList = !j9jitexceptiontable 0x000003FF41843AF8
        0x70: struct J9MemorySegment * classSegments = !j9memorysegment 0x000003FF4C0D1DB8
        0x78: struct J9RAMClassFreeListBlock * ramClassLargeBlockFreeList = !j9ramclassfreelistblock 0x000003FF4C0D45D0
        0x80: struct J9RAMClassFreeListBlock * ramClassSmallBlockFreeList = !j9ramclassfreelistblock 0x0000000000000000
        0x88: struct J9RAMClassFreeListBlock * ramClassTinyBlockFreeList = !j9ramclassfreelistblock 0x000003FF7408E3A8
        0x90: UDATA * ramClassUDATABlockFreeList = !j9x 0x000003FF4C0D43E8
        0x98: struct J9HashTable * redefinedClasses = !j9hashtable 0x0000000000000000
        0xa0: struct J9NativeLibrary * librariesHead = !j9nativelibrary 0x0000000000000000
        0xa8: struct J9NativeLibrary * librariesTail = !j9nativelibrary 0x0000000000000000
        0xb0: volatile UDATA gcRememberedSet = 0x0000000000000000 (0)
        0xb8: struct J9HashTable * moduleHashTable = !j9hashtable 0x0000000000000000
        0xc0: struct J9HashTable * packageHashTable = !j9hashtable 0x0000000000000000
        0xc8: struct J9HashTable * moduleExtraInfoHashTable = !j9hashtable 0x0000000000000000
        0xd0: struct J9HashTable * classLocationHashTable = !j9hashtable 0x0000000000000000
        0xd8: struct J9HashTable * classRelationshipsHashTable = !j9hashtable 0x0000000000000000
}
```

Here are also the flags with the printfs:
```
j> 14:48:50 Filip: counter **NOT** found = 245, ramClassFound = 0, loaderDead = 0, flags = 30, gcFlags = 0, class name = org/junit/runners/Suite
j> 14:48:50 Filip: counter **NOT** found = 246, ramClassFound = 0, loaderDead = 0, flags = 30, gcFlags = 0, class name = org/junit/runners/Suite
j> 14:48:50 Filip: counter **NOT** found = 247, ramClassFound = 0, loaderDead = 0, flags = 30, gcFlags = 0, class name = org/junit/runners/ParentRunner$3
j> 14:48:50 Filip: counter **NOT** found = 248, ramClassFound = 0, loaderDead = 0, flags = 30, gcFlags = 0, class name = org/junit/runners/ParentRunner$1
j> 14:48:50 Filip: counter **NOT** found = 249, ramClassFound = 0, loaderDead = 0, flags = 30, gcFlags = 0, class name = org/junit/runners/ParentRunner
j> 14:48:50 Filip: counter **NOT** found = 250, ramClassFound = 0, loaderDead = 0, flags = 30, gcFlags = 0, class name = org/junit/runners/ParentRunner
j> 14:48:50 Filip: counter **NOT** found = 251, ramClassFound = 0, loaderDead = 0, flags = 30, gcFlags = 0, class name = org/junit/runners/ParentRunner$2
j> 14:48:50 Filip: counter **NOT** found = 252, ramClassFound = 0, loaderDead = 0, flags = 30, gcFlags = 0, class name = org/junit/runners/ParentRunner
j> 14:48:50 Filip: counter **NOT** found = 253, ramClassFound = 0, loaderDead = 0, flags = 30, gcFlags = 0, class name = org/junit/runner/JUnitCore
j> 14:48:50 Filip: counter **NOT** found = 254, ramClassFound = 0, loaderDead = 0, flags = 30, gcFlags = 0, class name = org/junit/runner/JUnitCore
j> 14:48:50 Filip: counter **NOT** found = 255, ramClassFound = 0, loaderDead = 0, flags = 30, gcFlags = 0, class name = com/ibm/jtc/test/jltf/workunitadapters/JUnit4WorkUnit
```

They are always 0x1E for `flags` and 0 for `gcFlags` whenever we can't find the class via `peekClassHashTable`.

Reply:
@DanHeidinga any thoughts on the above? It seems that sometimes we are just unable to find the `J9Class*` while the `J9ROMMethod*` can be found.

Reply:
@fjeremic That would make sense if we had only loaded the J9ROMClass and hadn't yet created the J9Class but that can't happen for a case were we've been running the methods.  

If the class has been loaded, we should be able to find it in the classHashTable given the name and the right classloader.  

Does the `!classforname org/junit/runners/Suite` DDR extension find the class?  (Assuming that's the right name for missing class in the core).  And if it does, does its classloader match the one used in the `peekClassHashTable` call?

I think having a fallback for when the `J9Class` can't be found is necessary as the class may have been unloaded already but I'd like to ensure that the `peekClassHashTable` is working as expected and as effectively as it can

Reply:
@DanHeidinga so I re-ran the unit test and made it crash via `*(int*)(0) = 0` the first time we were unable to find the `J9Class*`. It spit out this:

```
j> 10:33:20 Filip: counter **NOT** found = 1, ramClassFound = 0, loaderDead = 0, flags = 30, gcFlags = 0, class name = com/ibm/test/decimals/TestDecimalData, classLoader = 0x3ff980705d8, romClass = 0x3ff66233bb8
```

Opening the core file in DDR and looking for the class yields the following:

```
> !classforname com/ibm/test/decimals/TestDecimalData
Searching for classes named 'com/ibm/test/decimals/TestDecimalData' in VM=3ff98013a70
!j9class 0x000003FF986C7A00 named com/ibm/test/decimals/TestDecimalData
Found 1 class(es) named com/ibm/test/decimals/TestDecimalData
```

Meaning that the class was found, however the classloader printed (and used in `peekClassHashTable`) is not the correct classloader which loaded the class:

```
> !j9class 0x000003FF986C7A00
J9Class at 0x3ff986c7a00 {
  Fields for J9Class:
        0x0: UDATA eyecatcher = 0x0000000099669966 (2573637990)
        0x8: struct J9ROMClass * romClass = !j9romclass 0x000003FF66233BB8
        0x10: struct J9Class ** superclasses = !j9x 0x000003FF986C77E8
        0x18: UDATA classDepthAndFlags = 0x00000000020E0001 (34471937)
        0x20: U32 classDepthWithFlags = 0x00000000 (0)
        0x24: U32 classFlags = 0x00000000 (0)
        0x28: struct J9ClassLoader * classLoader = !j9classloader 0x000003FF98070890

> !j9classloader 0x000003FF98070890
J9ClassLoader at 0x3ff98070890 {
  Fields for J9ClassLoader:
        0x0: struct J9Pool * sharedLibraries = !j9pool 0x0000000000000000
        0x8: struct J9HashTable * classHashTable = !j9hashtable 0x000003FF98259330
        0x10: struct J9HashTable * romClassOrphansHashTable = !j9hashtable 0x0000000000000000
        0x18: j9object_t classLoaderObject = !j9object 0x000003FF78070300 // sun/misc/Launcher$AppClassLoader
```

`sun/misc/Launcher$AppClassLoader` is the class loader which loaded the class we couldn't find. And the class loader we used in `peekClassHashTable` is:

```
> !j9classloader 0x3ff980705d8
J9ClassLoader at 0x3ff980705d8 {
  Fields for J9ClassLoader:
        0x0: struct J9Pool * sharedLibraries = !j9pool 0x000003FF98137A10
        0x8: struct J9HashTable * classHashTable = !j9hashtable 0x000003FF98082620
        0x10: struct J9HashTable * romClassOrphansHashTable = !j9hashtable 0x0000000000000000
        0x18: j9object_t classLoaderObject = !j9object 0x000003FF78035260 // com/ibm/oti/vm/BootstrapClassLoader
```

So the problem seems to be that `findROMClassFromPC` is not returning the correct classloader for the particular class. Any ideas as to why? Are we using this API correctly?

[1] https://github.com/eclipse/openj9/blob/b25d2d63793c8d12bba32f97c3848e4cb16409b7/runtime/vm/findmethod.c#L70-L90

Reply:
My current theory is that all ROM classes in the shared classes cache appear to be from the bootloader.  Working on proving that and then will propose a potential work around

Reply:
I just did a quick test and ran with `-Xshareclasses:none -Xnoaot` and the tests are passing now. We seem to no longer be encountering any `NULL` results from `peekClassHashTable`

Reply:
@pshipton helped me find the way we create segments for the SCC - code is here and assigning the bootloader to the scc segments: https://github.com/eclipse/openj9/blob/b25d2d63793c8d12bba32f97c3848e4cb16409b7/runtime/shared_common/CacheMap.cpp#L905